core/commands/ptp.go

```diff
@@ -0,0 +1,392 @@
+		cmds.StringArg("Peer", true, false, "Remote peer to connect to"),
+		cmds.StringArg("Protocol", true, false, "Protocol identifier."),
+		cmds.StringArg("BindAddress", false, false, "Address to listen for connection/s (default: /ip4/127.0.0.1/tcp/0)."),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		err = checkEnabled(n)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if !n.OnlineMode() {
+			res.SetError(errNotOnline, cmds.ErrClient)
+			return
+		}
+
+		addr, peer, err := ParsePeerParam(req.Arguments()[0])
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		proto := "/ptp/" + req.Arguments()[1]
+
+		bindAddr, _ := ma.NewMultiaddr("/ip4/127.0.0.1/tcp/0")
+		if len(req.Arguments()) == 3 {
+			bindAddr, err = ma.NewMultiaddr(req.Arguments()[2])
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+		}
+
+		listenerInfo, err := n.PTP.Dial(n.Context(), addr, peer, proto, bindAddr)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		output := PTPListenerInfoOutput{
+			Protocol: listenerInfo.Protocol,
+			Address:  listenerInfo.Address.String(),
+		}
+
+		res.SetOutput(&output)
+	},
+}
+
+var ptpCloseCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Closes an active p2p stream or listener.",
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("Identifier", false, false, "Stream HandlerID or p2p listener protocol"),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("all", "a", "Close all streams and listeners.").Default(false),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		err = checkEnabled(n)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if !n.OnlineMode() {
+			res.SetError(errNotOnline, cmds.ErrClient)
+			return
+		}
+
+		closeAll, _, _ := req.Option("all").Bool()
+
+		var proto string
+		var handlerID uint64
+
+		useHandlerID := false
+
+		if !closeAll {
+			if len(req.Arguments()) == 0 {
+				res.SetError(errors.New("no handlerID nor listener protocol specified"), cmds.ErrNormal)
+				return
+			}
+
+			handlerID, err = strconv.ParseUint(req.Arguments()[0], 10, 64)
+			if err != nil {
+				proto = "/ptp/" + req.Arguments()[0]
+			} else {
+				useHandlerID = true
+			}
+		}
+
+		if closeAll || useHandlerID {
+			for _, stream := range n.PTP.Streams.Streams {
+				if !closeAll && handlerID != stream.HandlerID {
+					continue
+				}
+				stream.Close()
+				if !closeAll {
+					break
+				}
+			}
+		}
+
+		if closeAll || !useHandlerID {
+			for _, listener := range n.PTP.Listeners.Listeners {
+				if !closeAll && listener.Protocol != proto {
+					continue
+				}
+				listener.Close()
+				if !closeAll {
+					break
+				}
+			}
+		}
```

Comment:
The logic here is quite complex, anyway to make it much simpler to read?

Probably it would be better to completly split off the close all, with handler, and with protocol into separate functions. Maybe even place them in the ptp package.